File: ios/Classes/Channels/Bridge.swift
import sdk_core_swift
import sdk_bletcpbridge_swift
import sdk_objectmodel_swift
import sdk_xyobleinterface_swift
import XyBleSdk

class XyoBridgeChannel: XyoNodeChannel {
    
  override
  init(registrar: FlutterPluginRegistrar, name: String) {
    super.init(registrar: registrar, name: name)
    BridgeManager.instance.restoreAndInitBridge()
    BridgeManager.instance.bridge.addListener(key: "BridgeChannel", listener: self)
    XYBluetoothManager.setup()
  }

  override func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    do {
      switch (call.method) {
      case "setArchivists":
        setArchivists(call, result: result)
        break
      case "getBlockCount":
        try getBlockCount(call, result: result)
        break
      case "getLastBlock":
        getLastBlock(call, result: result)
        break
      case "getBlocks":
        getBlocks(call, result: result)
        break
      case "selfSign":
        selfSign(call, result: result)
        break
      case "initiateBoundWitness":
        initiateBoundWitness(call, result: result)
        break
      default:
        super.handle(call, result: result)
      }
    } catch {
        result(FlutterError(code: "Error", message: "Unexpected Error", details:"\(error)"))
    }
  }

  override func onStart() -> String {
    XYBluetoothManager.start()
    status = XyoNodeChannel.STATUS_STARTED
    return status
  }

  override func onStop() -> String {
    XYBluetoothManager.stop()
    status = XyoNodeChannel.STATUS_STOPPED
    return status
  }

  private func setArchivists(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    result(FlutterMethodNotImplemented)
  }

  private func selfSign(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    do {
        try BridgeManager.instance.bridge.selfSignOriginChain()
    } catch {
        result(FlutterError(code: "Exception", message: "Excception", details:""))
    }

    //now that we have a new last block
    getLastBlock(call, result:result)
  }
  
  private func findDevice(_ deviceId: String) -> XyoBluetoothDevice? {
    var foundDevice: XyoBluetoothDevice?
    XYBluetoothDeviceFactory.devices.forEach { device in
      if (device.id == deviceId) {
        foundDevice = device as? XyoBluetoothDevice
      }
    }
    return foundDevice
  }
  
  private func initiateBoundWitness(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    
    guard
      let arguments = call.arguments as? Dictionary<String, Any>,
      let deviceId = arguments["deviceId"] as? String
    else {
      result(FlutterError(code: "Exception", message: "Excception", details:"DeviceId Invalid Type"))
      return
    }
    
    guard
      let device = findDevice(deviceId)
    else {
        result(FlutterError(code: "Exception", message: "Excception", details:"Device not Found"))
        return
    }
    
    BridgeManager.instance.bridge.collect(bleDevice: device)
    result(true)
  }

  private func getBlockCount(_ call: FlutterMethodCall, result: @escaping FlutterResult) throws {
    let hashes = BridgeManager.instance.blockRepo.getAllOriginBlockHashes()
    var hashObjects = [XyoObjectStructure]()
    
    let it = try! hashes.getNewIterator()
    
    while (try it.hasNext()) {
        let item = try it.next()
        hashObjects.append(item)
    }
    
    let models = hashesToBoundWitnesses(hashes: hashObjects)
    result(models.count)
  }

  private func getLastBlockData() -> DeviceBoundWitness? {
    guard let prevHash = BridgeManager.instance.stateRepo.getPreviousHash() else {
        return nil
    }
    
    
    let structure = XyoIterableStructure(value: prevHash.getBuffer())
    guard let hash = try? structure.get(index: 0) else {
        return nil
    }
    
    
    var boundWitnesses = hashesToBoundWitnesses(hashes: [hash])
    return boundWitnesses[0]
  }

  private func getLastBlock(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    let lastModel: DeviceBoundWitness? = getLastBlockData()
    result(try! lastModel?.serializedData())
  }
  
  private func getBlocksData() -> DeviceBoundWitnessList? {
    let hashes = BridgeManager.instance.blockRepo.getAllOriginBlockHashes()
    guard let iterator = try? hashes.getNewIterator() else {return nil}
    
    var list = [XyoObjectStructure]()
    
    while(try! iterator.hasNext()) {
      list.append(try! iterator.next())
    }
    
    let boundWitnesseses = hashesToBoundWitnesses(hashes: list)
    
    var bwList = DeviceBoundWitnessList()
    boundWitnesseses.forEach { item in
      bwList.boundWitnesses.append(item)
    }
    
    return bwList
  }
  
  private func getBlocks(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    let model: DeviceBoundWitnessList? = getBlocksData()
    result(try! model?.serializedData())
  }

  private func hashesToBoundWitnesses(hashes: [XyoObjectStructure]) -> [DeviceBoundWitness] {
    var models = [DeviceBoundWitness]()

    hashes.forEach { hash in
      let model = InteractionModel(hash.getBuffer().toByteArray(), date: Date(), linked: true)
      models.append(model.toBuffer)
    }

    return models
  }
}

extension XyoBridgeChannel: XyoNodeListener {
  public func onBoundWitnessStart() {}
  public func onBoundWitnessEndFailure() {}
  public func onBoundWitnessDiscovered(boundWitness: XyoBoundWitness) {}
  public func onBoundWitnessEndSuccess(boundWitness: XyoBoundWitness) {
    let lastBlock = getLastBlockData()!
    events.send(event: try! lastBlock.serializedData())
  }
}

File: android/src/main/kotlin/network/xyo/ble/xyo_ble/channels/bridge.kt
package network.xyo.ble.xyo_ble.channels

import android.content.Context
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.PluginRegistry
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.async
import kotlinx.coroutines.launch
import network.xyo.ble.devices.xy.XY4BluetoothDevice
import network.xyo.ble.flutter.protobuf.BoundWitness
import network.xyo.ble.generic.gatt.server.XYBluetoothAdvertiser
import network.xyo.ble.generic.gatt.server.XYBluetoothGattServer
import network.xyo.ble.generic.scanner.XYSmartScan
import network.xyo.ble.xyo_ble.InteractionModel
import network.xyo.modbluetoothkotlin.advertiser.XyoBluetoothAdvertiser
import network.xyo.modbluetoothkotlin.client.*
import network.xyo.modbluetoothkotlin.server.XyoBluetoothServer
import network.xyo.sdkcorekotlin.boundWitness.XyoBoundWitness
import network.xyo.sdkcorekotlin.node.XyoNodeListener
import network.xyo.sdkobjectmodelkotlin.structure.XyoIterableStructure
import network.xyo.sdkobjectmodelkotlin.structure.XyoObjectStructure
import java.util.*
import kotlin.collections.ArrayList

@kotlin.ExperimentalUnsignedTypes
class XyoBridgeChannel(context: Context, val smartScan: XYSmartScan, registrar: PluginRegistry.Registrar, name: String): XyoNodeChannel(context, registrar, name) {

    private val server = XYBluetoothGattServer(context.applicationContext)
    private val serverHelper = XyoBluetoothServer(server)
    private val xyAdvertiser = XYBluetoothAdvertiser(context)
    private lateinit var advertiser: XyoBluetoothAdvertiser

    init {
        GlobalScope.launch {
            bridgeManager.restoreAndInitBridge().await()
            bridgeManager.bridge.addListener("bridge", onBoundWitness)
        }

        XyoSentinelX.enable(true)
        XyoBridgeX.enable(true)
        XyoIosAppX.enable(true)
        XyoAndroidAppX.enable(true)
        XyoBluetoothClient.enable(true)
        XY4BluetoothDevice.enable(true)
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "setArchivists" -> setArchivists(call, result)
            "getBlockCount" -> getBlockCount(call, result)
            "getLastBlock" -> getLastBlock(call, result)
            "selfSign" -> selfSign(call, result)
            "getBlocks" -> getBlocks(call, result)
            "initiateBoundWitness" -> initiateBoundWitness(call, result)
            else -> super.onMethodCall(call, result)
        }
    }

    private fun createNewAdvertiser(): XyoBluetoothAdvertiser {
        return XyoBluetoothAdvertiser(
                Random().nextInt(Short.MAX_VALUE + 1).toUShort(),
                Random().nextInt(Short.MAX_VALUE + 1).toUShort(),
                xyAdvertiser)
    }

    override suspend fun onStart(): XyoNodeChannelStatus {
        smartScan.addListener("bridge", bridgeManager.bridge.scanCallback)
        serverHelper.listener = bridgeManager.bridge.serverCallback
        server.startServer()
        serverHelper.initServer()
        advertiser = createNewAdvertiser()
        advertiser.configureAdvertiser()
        advertiser.startAdvertiser()
        smartScan.start()
        status = XyoNodeChannelStatus.Started
        return status
    }

    override suspend fun onStop(): XyoNodeChannelStatus {
        smartScan.removeListener("bridge")
        serverHelper.listener = null
        server.stopServer()
        status = XyoNodeChannelStatus.Stopped
        return status
    }

    private fun setArchivists(call: MethodCall, result: MethodChannel.Result) = GlobalScope.launch {
        notImplemented(result)
    }

    private fun selfSign(call: MethodCall, result: MethodChannel.Result) = GlobalScope.launch {
        bridgeManager.bridge.selfSignOriginChain().await()

        //now that we have a new last block
        getLastBlock(call, result)
    }

    private fun initiateBoundWitness(call: MethodCall, result: MethodChannel.Result) = GlobalScope.launch {
        var didInitiate = false
        val deviceId = (call.arguments as? Map<String, Any?>)?.get("deviceId") as? String
        deviceId?.let {
            val device = smartScan.deviceFromId(it) as? XyoBluetoothClient
            device?.let {
                bridgeManager.bridge.tryBoundWitnessWithDevice(device)
                didInitiate = true
            }
        }
        if (didInitiate) {
            sendResult(result, true)
        } else {
            sendError(result, "Not Found", "Device not found for connection")
        }
    }

    private fun getBlockCount(call: MethodCall, result: MethodChannel.Result) = GlobalScope.launch {
        val hashes = bridgeManager.bridge.blockRepository.getAllOriginBlockHashes().await() ?: return@launch sendError(result, "FAILED")
        val models = hashesToBoundWitnesses(hashes)

        sendResult(result, models?.count())
    }

    private fun getLastBlockData() = GlobalScope.async {
        val hash = bridgeManager.bridge.originState.previousHash ?: return@async null
        val structure = XyoIterableStructure(hash.bytesCopy, 0)

        val hashArray = arrayOf(structure[0])
        val hashArrayIterator = hashArray.iterator()
        val boundWitnesses = hashesToBoundWitnesses(hashArrayIterator)
        return@async boundWitnesses?.first()
    }

    private fun getLastBlock(call: MethodCall, result: MethodChannel.Result) = GlobalScope.launch {
        val lastModel: BoundWitness.DeviceBoundWitness? = getLastBlockData().await()
        sendResult(result, lastModel?.toByteArray())
    }

    private fun getBlocksData() = GlobalScope.async {
        val hashes = bridgeManager.blockRepo.getAllOriginBlockHashes().await()
        val iterator = hashes?.iterator()
        return@async hashesToBoundWitnesses(iterator)
    }

    private fun getBlocks(call: MethodCall, result: MethodChannel.Result) = GlobalScope.launch {
        val blocks = getBlocksData().await()
        val builder = BoundWitness.DeviceBoundWitnessList.newBuilder()
        builder.addAllBoundWitnesses(blocks?.toList())
        sendResult(result, builder.build().toByteArray())
    }

    private fun hashesToBoundWitnesses(hashes: Iterator<XyoObjectStructure>?): ArrayList<BoundWitness.DeviceBoundWitness>? {
        if (hashes == null) return null;
        val models = ArrayList<BoundWitness.DeviceBoundWitness>()

        for (hash in hashes) {
            val model = InteractionModel(bridgeManager.bridge.blockRepository, hash.bytesCopy, Date(), true)
            models.add(model.toBuffer())
        }

        return models
    }

    private fun notifyNewBoundWitness() {
        GlobalScope.launch {
            val boundWitness = getLastBlockData().await()

            if (boundWitness != null) {
                events.send(boundWitness.toByteArray())
            }
        }
    }

    private val onBoundWitness = object : XyoNodeListener() {
        override fun onBoundWitnessEndSuccess(boundWitness: XyoBoundWitness) {
            notifyNewBoundWitness()
        }
    }
}